sequencer/src/state.rs

```diff
@@ -157,17 +172,21 @@ impl ValidatedState {
                 Some(updated)
             }
         })?;
-        // Check if we were unable to do the update because the required Merkle path is missing.
-        ensure!(
-            res.expect_not_in_memory().is_err(),
-            format!("missing account state for {account}")
-        );
-        // Fail if there was an error during `update_with`, otherwise succeed.
+
+        // Fail if there was an error during `persistent_update_with` (e.g. insufficient balance).
         if let Some(err) = err {
-            Err(err)
-        } else {
-            Ok(())
+            return Err(err);
```

Comment:
Not really an issue w/ this PR, but I wonder if `persistent_update_with` would take a closure that returns a `Result` instead of an `Option`. I *think* we could then bubble the error up to the caller instead of mutating `err`. This code *appears* to demonstrate that we do need to handle more than just the `Some` and `None` cases. I might be confused though.

Reply:
Yeah I think that would be cleaner. This pattern is basically just implementing that in a messier way. @mrain maybe you can clean this up in jf-merkle-tree whenever you find yourself with some spare time?